### ISSUE TITLE
Force Layout Editor to Show Graph at Highest Rank

### ISSUE THREAD

**Issue by mitchwagner** [Murali-group/GraphSpace#305]
The layout editor presently will manipulate nodes outside of a current rank during editing. For example, the following is a possible workflow: 

1. User views graph of blue and green nodes at rank 5, and sees 2 nodes of each color.
2. User select blue and green nodes and manipulates them into a square using a layout button, and notices some empty spaces in the square.
3. User exits the layout editor, and is prompted to save.
4. User now changes rank to 10.
5. User sees that more blue and green nodes have appeared in the square, filling the empty spaces noticed before.

From a user perspective, this does not feel very intuitive. I would expect only the nodes I selected at the particular rank to be manipulated. At the very least, it feels like there is friction between exiting the layout editor, not saving the layout, and changing the rank.

Furthermore, the present implementation runs the risk of a user thinking they have laid out a graph completely, only to realize that the rank slider was not as high as possible.

**Comment by jlaw9** [comment]
Another perspective is a user posts that 1000 paths, and then only wants to layout the top 50. In that case, it would be much better to be able to use the automated layouts and layout editor on only the visible nodes.

**Comment by mitchwagner** [comment]
@jlaw9 that is a very  good point, but this is still complicated by the fact that the layout manager currently takes those unseen nodes into account and will leave holes in the square or circles it creates, which limits this functionality.

I can post images as examples soon.

**Comment by jlaw9** [comment]
@mitchwagner yes you're right. Forcing the layout editor to show all of the nodes and edges is better than the current functionality, but my preferred solution is to fix the layout manager so that the hidden nodes are not included in layout. If we removed the filtered nodes and edges instead of hiding them, the layout manager would operate correctly. That would require that we store two copies of the graph. One that has all of the data, and another for visualizing with the filters. I guess I should open a separate issue for this.

**Comment by jahandaniyal** [comment]
@adbharadwaj I see this issue has not been assigned yet. Can I take this issue ?

**Comment by jahandaniyal** [comment]
@mitchwagner @jlaw9 Did you envision something like this ?

![bug 334](https://user-images.githubusercontent.com/4581090/36301682-9959d5c4-12fd-11e8-88f5-0b4b383128fb.gif)

**Comment by jlaw9** [comment]
@jahandaniyal Looks great! This is the functionality I was expecting. Also somewhat related to this issue is #257 (if you save the layout with a filter/rank, the filter setting should also be saved).

**Comment by jahandaniyal** [comment]
@jlaw9 Thanks. I will check #257.
@adbharadwaj Can I work on this issue? I think I might have a solution (requires minimal changes).

**Comment by jlaw9** [comment]
Another aspect of this issue is the automatic layouts. If set the rank filter to the top 20 paths, I would expect an automatic layout such as Spring Embedder or Grid to be applied only on the nodes currently seen (i.e., not filtered out). 

@adbharadwaj what do you think?

Also, @jahandaniyal did you test the "Arrange nodes" tools in the layout editor?

**Comment by tmmurali** [comment]
@adbharadwaj @mitchwagner @jlaw9 @jahandaniyal 

There are at least two related issues on the layout editor. We have to consider all of them together and come up with a solution where the semantics of the operations are consistent across different scenarios.

My worry is that a discussion broken up across different issues on GitHub is going to lead to a messy solution. Can we create a Wiki page that summarises the issues and proposes coherent solutions? If Wiki is not the best way forward, I am open to other suggestions.

**Comment by jahandaniyal** [comment]
@jlaw9 Its possible! The `cy.layout()` function also works on a subset of the Graph. `cy.layout({graph_Subset}) `will apply layout only to the subset and, since it returns a layout object we could store its reference to track changes to the graph layout (Git for graphs).
[http://js.cytoscape.org/#eles.layout](url)

All of the layouts work fine except Cola. Cola is only working if we select the entire graph i.e if we set slider to Maximum. 
 
![soln 305_2](https://user-images.githubusercontent.com/4581090/36351134-034649fe-149d-11e8-8766-f2a263c4fccb.gif)

Prof. @tmmurali thank you for your suggestions regarding the issue. I think Wiki pages or something like JIRA will be very helpful for the new contributors.

**Comment by adbharadwaj** [comment]
@tmmurali Usually open-source projects close the related issue after tagging them as duplicate in the parent issue. Here is the documentation by GitHub on how to do it - [About duplicate issues and pull requests](https://help.github.com/articles/about-duplicate-issues-and-pull-requests/). Another common strategy is to add a duplicate label so that contributors can quickly identify an issue as duplicate without scrolling through the thread.

**Comment by tmmurali** [comment]
Let us take both steps, as seems to the case in the link you included.
